ERROR: Failed building wheel for astropy when attempting to build from source

### Answer

https://github.com/astropy/astropy/issues/15059#issuecomment-1633272304

### Description

I am trying to follow the instructions for "Building from Source" at this link:
https://docs.astropy.org/en/stable/install.html



### Expected behavior

I would expect astropy to install and be usable in the same way as if I installed using `pip3 install astropy`.  It is worth noting that `pip3 install astropy` works as expected. When I try to install per the astropy docs, I get a few pages of errors. The bits of the error message that seem most likely to be useful are copied below:
```

Building wheels for collected packages: astropy
  Building wheel for astropy (PEP 517) ... error
  ERROR: Command errored out with exit status 1:
   command: /usr/local/bin/python3 /usr/local/lib/python3.9/site-packages/pip/_vendor/pep517/in_process/_in_process.py build_wheel /tmp/tmp8o9qty2h
       cwd: /tmp/pip-req-build-diyrb7ti
  Complete output (1661 lines):

<cropped>


  astropy/wcs/src/wcslib_celprm_wrap.c: In function 'PyCelprm_deepcopy':
  astropy/wcs/src/wcslib_celprm_wrap.c:167:22: warning: initialization from incompatible pointer type [enabled by default]
       PyCelprm* copy = PyCelprm_new(&PyCelprmType, NULL, NULL);
                        ^
  astropy/wcs/src/wcslib_celprm_wrap.c: In function 'PyCelprm_set_ref':
  astropy/wcs/src/wcslib_celprm_wrap.c:327:5: warning: passing argument 1 of 'PyArray_DIMS' from incompatible pointer type [enabled by default]
       size = PyArray_SIZE(value_array);
       ^

<lots of other similar error messages edited out>

  gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -I./astropy/io/fits/_tiled_compression/src -I/usr/local/include/python3.9 -c ./astropy/io/fits/_tiled_compression/src/compression.c -o build/temp.linux-x86_64-cpython-39/./astropy/io/fits/_tiled_compression/src/compression.o
  ./astropy/io/fits/_tiled_compression/src/compression.c: In function 'compress_plio_1_c':
  ./astropy/io/fits/_tiled_compression/src/compression.c:117:3: error: 'for' loop initial declarations are only allowed in C99 mode
     for (int ii = 0; ii < tilesize; ii++)  {
     ^
  ./astropy/io/fits/_tiled_compression/src/compression.c:117:3: note: use option -std=c99 or -std=gnu99 to compile your code
  error: command '/usr/bin/gcc' failed with exit code 1
  ----------------------------------------
  ERROR: Failed building wheel for astropy
Failed to build astropy
ERROR: Could not build wheels for astropy which use PEP 517 and cannot be installed directly
```

### How to Reproduce

I am running Centos7 with updated kernel Linux 5.18.7-1.el7.elrepo.x86_64 x86_64.

Relevant bits from my history:
```
sudo yum install python3-devel python3-numpy python3-setuptools python3-Cython python3-pytest-astropy
git clone https://github.com/astropy/astropy.git
cd astropy/
python3 -m pip install .
```

I get the same result if I replace the last line above with
`pip3 install .`

### Versions
```
>>> import astropy; print("astropy", astropy.__version__)
/generator/tmp/astropy/astropy/version.py:12: UserWarning: could not determine astropy package version; this indicates a broken installation
  warnings.warn(
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/generator/tmp/astropy/astropy/__init__.py", line 176, in <module>
    _initialize_astropy()
  File "/generator/tmp/astropy/astropy/__init__.py", line 150, in _initialize_astropy
    raise ImportError(
ImportError: You appear to be trying to import astropy from within a source checkout or from an editable installation without building the extension modules first. Either run:

  pip install -e .

or

  python setup.py build_ext --inplace

to make sure the extension modules are built
>>> import numpy; print("Numpy", numpy.__version__)
Numpy 1.25.0
>>> import erfa; print("pyerfa", erfa.__version__)
pyerfa 2.0.0.3
>>> import scipy; print("Scipy", scipy.__version__)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'scipy'
>>> import matplotlib; print("Matplotlib", matplotlib.__version__)
Matplotlib 3.7.1
```

--- a/docs/install.rst
+++ b/docs/install.rst
@@ -345,12 +345,12 @@ installing with::
 
 which will install into a default directory in your home directory.
 
-If you get an error directing use of option -std=c99 or -std=gnu99, you can try
+If you get an error directing use of option ``-std=c99`` or ``-std=gnu99``, you can try
 installing with::
 
     CFLAGS='-std=c99' pip install -e .
 
-This is necessary for use with Centos7.
+This is necessary for use with CentOS7.
 
 .. _external_c_libraries:
 

--- a/docs/install.rst
+++ b/docs/install.rst
@@ -345,6 +345,13 @@ installing with::
 
 which will install into a default directory in your home directory.
 
+If you get an error directing use of option -std=c99 or -std=gnu99, you can try
+installing with::
+
+    CFLAGS='-std=c99' pip install -e .
+
+This is necessary for use with Centos7.
+
 .. _external_c_libraries:
 
 External C Libraries

